src/object.jl

```diff
@@ -372,3 +372,198 @@ function Base.read(io::PrefetchedDownloadStream, ::Type{UInt8})
     end
     return b
 end
+
+function _upload_task(io; kw...)
+    try
+        (part_n, upload_buffer) = take!(io.upload_queue)
+        # upload the part
+        parteTag, wb = uploadPart(io.store, io.url, upload_buffer, part_n, io.uploadState; io.credentials, kw...)
+        Base.release(io.sem)
+        # add part eTag to our collection of eTags
+        Base.@lock io.cond_wait begin
+            if length(io.eTags) < part_n
+                resize!(io.eTags, part_n)
+            end
+            io.eTags[part_n] = parteTag
+            io.ntasks -= 1
+            notify(io.cond_wait)
+        end
+    catch e
+        isopen(io.upload_queue) && close(io.upload_queue, e)
+        Base.@lock io.cond_wait begin
+            io.exc = e
+            notify(io.cond_wait, e, all=true, error=true)
+        end
+    end
+    return nothing
+end
+
+"""
+    MultipartUploadStream <: IO
+    MultipartUploadStream(args...; kwargs...) -> MultipartUploadStream
+
+An in-memory IO stream that uploads chunks to a URL in blob storage.
+
+Data chunks are written to a channel and spawned tasks read buffers from this channel.
+We expect the chunks to be written in order.
```

Comment:
I'd mention this is currently spawns one task per chunk and lets explicitly mention the `write` method and also `put` as an alternative.

Reply:
Could you please elaborate a bit more on the use of `put` as an alternative here? Doesn't `put` require the total input to be known upfront?

Reply:
Just to mention it, that if you don't need to upload files by part (or if you simply cannot, because your file is too small), you can always use `put`